framework/src/problems/SubProblem.C

```diff
@@ -67,49 +68,138 @@ SubProblem::SubProblem(const InputParameters & parameters)
 SubProblem::~SubProblem() {}
 
 TagID
-SubProblem::addVectorTag(TagName tag_name)
+SubProblem::addVectorTag(const TagName & tag_name,
+                         const Moose::VectorTagType type /* = Moose::VECTOR_TAG_RESIDUAL */)
 {
-  auto tag_name_upper = MooseUtils::toUpper(tag_name);
-  auto existing_tag = _vector_tag_name_to_tag_id.find(tag_name_upper);
-  if (existing_tag == _vector_tag_name_to_tag_id.end())
-  {
-    auto tag_id = _vector_tag_name_to_tag_id.size();
+  if (type == Moose::VECTOR_TAG_ANY)
+    mooseError("Vector tag type cannot be VECTOR_TAG_ANY");
 
-    _vector_tag_name_to_tag_id[tag_name_upper] = tag_id;
+  const auto tag_name_upper = MooseUtils::toUpper(tag_name);
 
-    _vector_tag_id_to_tag_name[tag_id] = tag_name_upper;
+  // First, see if the tag exists already
+  for (const auto & vector_tag : _vector_tags)
+  {
+    mooseAssert(_vector_tags[vector_tag._id] == vector_tag, "Vector tags index mismatch");
+    if (vector_tag._name == tag_name_upper)
+    {
+      if (vector_tag._type != type)
+        mooseError("While attempting to add vector tag with name '",
+                   tag_name_upper,
+                   "' and type ",
+                   type,
+                   ",\na tag with the same name but type ",
+                   vector_tag._type,
+                   " was found.\n\nA tag can only exist with one type.");
+
+      return vector_tag._id;
+    }
   }
 
-  return _vector_tag_name_to_tag_id.at(tag_name_upper);
+  // Doesn't exist - create it
+  const TagID new_tag_id = _vector_tags.size();
+  // Primary storage for all tags where the index in the vector == the tag ID
+  _vector_tags.emplace_back(new_tag_id, tag_name_upper, type);
+  // Secondary storage for each type so that we can have quick access to all tags of a type
+  _typed_vector_tags[type].emplace_back(new_tag_id, tag_name_upper, type);
+
+  verifyVectorTags();
+
+  return new_tag_id;
 }
 
 bool
-SubProblem::vectorTagExists(const TagName & tag_name)
+SubProblem::vectorTagExists(const TagName & tag_name) const
 {
-  auto tag_name_upper = MooseUtils::toUpper(tag_name);
+  mooseAssert(verifyVectorTags(), "Vector tag storage invalid");
```

Comment:
Yes, I still like the original O(log(n)) algorithm that is better than O(n). This runtime will be used often.

Reply:
We need  a map here